### PR TITLE
Fix 147

### PR DESCRIPTION
**Pull request by AMairesse**
Add support for multi-chars labels.
Propose a completed labels.json with possible English double or triple-chars.

### CODE REVIEW

**Comment by AMairesse** [comment]
Hi,
This is a proposed implementation for #147.
I've run some training on AN4 but I can't see any improvement, I'm even not able to obtain the 4.88 CER from "pre-trained model v2".
I was not able to do a training on LibriSpeech because it seems that there is a memory leak in pytorch for the cuda part and my training end up stuck before completing the first epoch.
Maybe you should give it a try before merging, in order to confirm if it's useful. AN4 isn't a representative dataset for this modification of the labels.

I think "'nt" label for example could improve because it's usually associated with a specific pronunciation.
See https://arxiv.org/abs/1609.05935v2 : 

> To identify repeated letters, we use special double-letter units to represent repeated characters like "ll".
> Finally, to improve the readability of the output without any further processing, we attach apostrophes to the following letters, creating units like the “’d” in “we’d.”

So that should be useful.

**Comment by AMairesse** [comment]
I've been doing some tests on LibriSpeech and the result is pretty good with this PR.
Launching a training with the command line from the pre-trained models V2 :

- without it and only one character long labels (current master) : 
Test-clean : Average WER 15.966	Average CER 4.308
Test-other : Average WER 37.321	Average CER 13.137

- with multi-characters labels (master + this PR) :
Test-clean : Average WER 12.675	Average CER 3.482	
Test-other : Average WER 33.102	Average CER 11.637

Note : I wasn't able to train if the first epoch is done on a size ordered dataset so I forced a shuffle even for the first epoch in the two tests (see #172).

Also I wasn't able to get a CER of 3.09 on test-clean like there is on the pre-trained models. Does the pre-trained models really had only 15 epochs ? Am I missing something here ?

**Comment by SeanNaren** [comment]
hi @AMairesse thanks for your work! Could you show me a use case where this is useful? It confuses the parse transcript function a bit and want to make sure this makes sense to add

**Comment by AMairesse** [comment]
I agree that the parse transcript function is quite more complex, I wasn't able to find a better way to do it. The idea is to be able to parse a text with labels of any size, starting from the largest labels from labels.json because we want for example "ll" to be encoded as the label for "ll" and not as two "l" labels.

This idea of using special labels for double letters and apostrophes and letters come from a research paper from 2016 : https://arxiv.org/abs/1609.05935v2
The idea is that CTC being based on collapsing same letters when they follow each other, learning a double letter require the model to learn to include a blank label. For example "LLLLLLL" would be "L" and "LLLL_LLLLL" would be "LL". Including special labels for double letters should make the model more precise and decoding easier.
In the paper they evaluate it but combined with another trick  which is to use a capital letter at the start of each word and remove the need for a label for the space. They get an improvement on their model when using these two tricks (see part 4.1 and 4.2 and table 1 in the paper).

I implemented only the first part which is to have specials labels for double letters and specific english usage of apostrophe. With this I get an significant improvement on LibriSpeech.
I think the better result come from less errors on double letters on average.

**Comment by ryanleary** [comment]
I'm looking over the implementation to ensure I fully grok how this is implemented. It may be somewhat nontrivial because if the correct output is "seriaLLy", and the model outputs "L___LLLL", that shouldn't get penalized in the loss function even though there is an "LL" class.

@AMairesse have you read the gram-ctc paper? https://arxiv.org/abs/1703.00096

**Comment by AMairesse** [comment]
@ryanleary : thanks for pointing out the paper, never read it before. It's clear that having a gram-ctc would be way better than the trick I'm proposing here :-)
My patch is far from such a solution and I agree that looking into gram-CTC would be a better solution : more efficient and language agnostic.

A quick search into Github gave me : https://github.com/musyoku/chainer-gram-ctc
It's quite over my current knowledge of the matter... I've used CTC as a black box until today, understanding what it was doing but not how. I'll try to look into it in the coming weeks and see if it can be used "as it". If it is maybe it could be a good start and could be built upon.

**Comment by miguelvr** [comment]
isn't it much easier to save the labels as a list instead of a string? that should do the trick
